design-proposals/vm-rollout-strategy/vm-rollout-strategy.md

```diff
@@ -0,0 +1,195 @@
+# Overview
+
+Today changes to a VM object (i.e. changing VM.spec.template) get applied, but only take effect upon the next restart - we call these pending changes “staged”. This is a proposal to introduce a rollout strategy field to control how changes to the VM object will propagate to the VirtualMachineInstance object.
+
+
+## Motivation
+
+In most cases, the user would prefer the changes to the VM object to immediately propagate to the VMI object, as long as the rollout does not interrupt the workload. This behavior would be similar to how the Deployments roll out DaemonSets.
+
+
+## Goals
+
+-   Provide a way to propagate changes from a running VM object to its corresponding VMI immediately without interrupting the workload.
+    
+-   Maintain backwards compatibility
+    
+
+## Non Goals
+
+-   Successfully apply all the changes to the Virtual Machine Instance Object. Not all fields can be propagated to the guest immediately. Some fields should remain staged, such as those that affect the guest ABI (e.g. machine type)
+    
+-   Make LiveUpdate the default. We defer this decision.
+    
+
+## Definition of Users
+
+A user - a VM owner - with (at a minimum) create, update, and patch access for VM objects within a namespace.
+
+
+## User Stories
+
+As a user I want changes that I applied to the VM object to propagate to the VMI immediately if possible.
+
+As a user, I want to be notified about changes that require a restart of the VM.
+
+## Repos
+
+kubevirt/kubevirt
+
+
+# Design
+
+A new field called rolloutStrategy will be introduced in the VM spec. This field will control the behavior of the VM object. Two possible values will be available:
+
+1.  Setting RolloutStrategy to `LiveUpdate` means all changes to the VM object will be immediately applied to the VMI.
+    
+
+A condition called `RestartRequired` will be added to the VM if such changes are not possible to propagate,
```

Comment:
If someone changes a configuration (e.g. `maxSockets`) with a VM using the `stage` strategy, then the condition is being added, will the condition be removed by reverting the configuration?

As a VM owner using a `staged` strategy, I think it's valuable to rely on this condition to check whether a configuration gets applied immediately or not.

In other words, VM owners could use the following approach if they wish to update a configuration:
1. Change a configuration
2. Check whether the condition is added

I think of a situation where someone wants to apply the config immediately if possible, but if he can't, then he doesn't want to stage it (one reason could be that he would be surprised by the changes in the next reboot and prefers to explicitly modify these configs only when the VM is shut down).

Not sure if I'm overthinking here.

Reply:
I suppose this evaluation would be part of the VM controller. It would add the condition when needed and remove it when the change has been applied or the strategy changed.
Similar to what we do with the live migration condition.